### PR TITLE
Add curl advisories importer

### PR DESCRIPTION
**Pull request by ambuj-1211**
This pr resolves the issue #1166.
The advisories were added properly and the site was working fine. I also registered the importer in the __init__.py file.
This image shows the curl advisories in vulnerablecode

### CODE REVIEW

**Comment by ambuj-1211** [comment]
@TG1999 please let me know if there are any further changes and everything is correct or not.

File: vulnerabilities/tests/test_curl.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# from importer import AdvisoryData
+from univers.versions import SemverVersion
+from vulnerabilities.importers import curl
+from vulnerabilities.tests import util_tests
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+TEST_DATA = os.path.join(BASE_DIR, "test_data/curl")
+
+
+def test_is_semver_version():
```
Comment (ziadhany) [inline vulnerabilities/tests/test_curl.py:26]: Fetching the URL inside the test is a bad idea. 
it makes tests depend on internet connectivity.
try to use a static file instead and please add a test to the importer 
ex:
https://github.com/nexB/vulnerablecode/pull/897/files#diff-1c1d12198c8d593ef053851512de82d7ce743fa976dfe870777ca9032db08606R21-R36

File: vulnerabilities/importers/curl.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                logger.error(f"Invalid CVE ID: {cve_id} in package {package}")
+                continue
+            yield parse_advisory_data(data)
+
+
+def parse_advisory_data(raw_data) -> AdvisoryData:
+
+    d1 = get_item(raw_data, "affected")[0] if len(get_item(raw_data, "affected")) > 0 else []
+    d2 = get_item(d1, "ranges")[0] if len(get_item(d1, "ranges")) > 0 else []
+    d3 = get_item(d2, "events")[1] if len(get_item(d2, "events")) > 1 else {}
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:54]: Please try to use meaningful variable names

**Review by ziadhany — CHANGES_REQUESTED** [review]
This is a good start but try to make use of all available data ex: `details, CWE , ..`

**Comment by ambuj-1211** [comment]
on it

**Comment by ambuj-1211** [comment]
@ziadhany @TG1999 Done with the changes as mentioned.

File: vulnerabilities/importers/curl.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        get_item(raw_data, "affected")[0] if len(get_item(raw_data, "affected")) > 0 else []
+    )
+    dictionary2 = (
+        get_item(dictionary1, "ranges")[0] if len(get_item(dictionary1, "ranges")) > 0 else []
+    )
+    dictionary3 = (
+        get_item(dictionary2, "events")[1] if len(get_item(dictionary2, "events")) > 1 else {}
+    )
+
+    fixed_version = SemverVersion(dictionary3.get("fixed") or "")
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:71]: you need to check if the `type = "SEMVER"`  before decide if this version is a SemverVersion

File: vulnerabilities/importers/curl.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    purl = PackageURL(type="generic", namespace="curl.se", name="curl")
+    affected_version_range = GenericVersionRange.from_versions(
+        raw_data.get("affected")[0].get("versions") or []
+    )
+
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:76]: it would be better if we use two rows instead of one.

File: vulnerabilities/importers/curl.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    dictionary1 = (
+        get_item(raw_data, "affected")[0] if len(get_item(raw_data, "affected")) > 0 else []
+    )
+    dictionary2 = (
+        get_item(dictionary1, "ranges")[0] if len(get_item(dictionary1, "ranges")) > 0 else []
+    )
+    dictionary3 = (
+        get_item(dictionary2, "events")[1] if len(get_item(dictionary2, "events")) > 1 else {}
+    )
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:69]: Please create meaningful variables names

File: vulnerabilities/importers/curl.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        summary=raw_data.get("summary") or "",
+        affected_packages=[affected_package],
+        references=references,
+        date_published=date_published,
+        weaknesses=weaknesses,
+        url=raw_data.get("database_specific", {}).get("URL", ""),
+    )
+
+
+def get_cwe_from_curl_advisory(raw_data):
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:104]: please add a doctest

File: vulnerabilities/importers/curl.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            if not cve_id.startswith("CVE"):
+                package = data.get("database_specific").get("package")
+                logger.error(f"Invalid CVE ID: {cve_id} in package {package}")
+                continue
+            yield parse_advisory_data(data)
+
+
+def parse_advisory_data(raw_data) -> AdvisoryData:
+
+    dictionary1 = (
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:61]:
```suggestion
    affected = (
```

File: vulnerabilities/importers/curl.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        get_item(raw_data, "affected")[0] if len(get_item(raw_data, "affected")) > 0 else []
+    )
+    dictionary2 = (
+        get_item(dictionary1, "ranges")[0] if len(get_item(dictionary1, "ranges")) > 0 else []
+    )
+    dictionary3 = (
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:67]:
```suggestion
    events = (
```

File: vulnerabilities/importers/curl.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    affected_package = AffectedPackage(
+        package=purl, affected_version_range=affected_version_range, fixed_version=fixed_version
+    )
+
+    database_specific = raw_data.get("database_specific") or {}
+    severity = VulnerabilitySeverity(
+        system=SCORING_SYSTEMS["generic_textual"], value=database_specific.get("severity", "")
+    )
+
+    references = [Reference(url=database_specific.get("www") or "", severities=[severity])]
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:86]: I think we need to have two References  ( `www`, `issue` )

```suggestion
    references = []
    ref_www = database_specific.get("www") or ""
    ref_issue = database_specific.get("issue") or ""
    if ref_www:
        references.append(Reference(url=ref_www, severities=[severity]))
    if ref_issue:
        references.append(Reference(url=ref_issue))
```

**Review by ziadhany — CHANGES_REQUESTED** [review]
@Abhigyankrsingh Nice work! just a few nits for your consideration.

**Comment by ziadhany** [comment]
@ambuj-1211 please add a curl improver to the `valid_versions.py` 

ex: https://github.com/nexB/vulnerablecode/blob/main/vulnerabilities/improvers/valid_versions.py#L472

**Comment by ambuj-1211** [comment]
> work! just a few nits for your consi

I think you tagged someone else in this comment, is it a mistake or something else?

**Comment by ziadhany** [comment]
> work! just a few nits for your consi
> 
> I think you tagged someone else in this comment, is it a mistake?
> And I am working on the changes now.

sorry it was a mistake and I fixed it

File: vulnerabilities/importers/curl.py (hunk 6 above)
Comment (ambuj-1211) [reply vulnerabilities/importers/curl.py:104]: do I need to add a doctest for all the functions or just for cwe and parse advisory data function

Comment (ziadhany) [reply vulnerabilities/importers/curl.py:104]: you should add a doctest to increase function readability.
I think you should add a doctest to `get_cwe_from_curl_advisory` and add just docs to `parse_advisory_data`

**Comment by ambuj-1211** [comment]
@ziadhany please review the changes and do let me know.

File: vulnerabilities/importers/curl.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    """
+    affected = get_item(raw_data, "affected")[0] if len(get_item(raw_data, "affected")) > 0 else []
+
+    ranges = get_item(affected, "ranges")[0] if len(get_item(affected, "ranges")) > 0 else []
+    events = get_item(ranges, "events")[1] if len(get_item(ranges, "events")) > 1 else {}
+    type_of_version = get_item(ranges, "type") if get_item(ranges, "type") else ""
+    # fixed_version = ""
+    if type_of_version=="SEMVER":
+        fixed_version = SemverVersion(events.get("fixed") or "")
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:111]: This seems like a cleaner way to do it.
```suggestion
    version_type = get_item(ranges, "type") if get_item(ranges, "type") else ""
    fixed_version = events.get("fixed")
    if version_type=="SEMVER" and fixed_version:
        fixed_version = SemverVersion(fixed_version)
```

**Review by ziadhany — CHANGES_REQUESTED** [review]
Great work! please fix those failing tests then squash the commits

**Comment by ambuj-1211** [comment]
@ziadhany I made the changes, should I also update my branch?

**Comment by ziadhany** [comment]
@ambuj-1211 to fix the test update the `vulnerabilities/tests/test_data/curl/expected-curl-advisory-output.json`
and it would be good if you add more test examples :
`vulnerabilities/tests/test_data/curl/expected-curl-advisory-output1.json`
`vulnerabilities/tests/test_data/curl/expected-curl-advisory-output2.json`

**Comment by ambuj-1211** [comment]
@ziadhany @TG1999 i have made all the required changes and squashed all the commits to make the history clean, moreover I also had updated my branch according to the main branch.
Please review it.

**Comment by ambuj-1211** [comment]
@ziadhany some checks are not successfull but they are because of openssl url do I need to change it to some thing, I am asking because it is not related to curl advisories.

**Comment by ziadhany** [comment]
> @ziadhany some checks are not successfull but they are because of openssl url do I need to change it to some thing, I am asking because it is not related to curl advisories.

No, this is the docs test ( not related to this pull request )

**Comment by ambuj-1211** [comment]
so this pr is good to go?

File: vulnerabilities/importers/curl.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+logger = logging.getLogger(__name__)
+
+
+class CurlImporter(Importer):
+
+    spdx_license_expression = "MIT"
+    license_url = "https://github.com/curl/curl-www/blob/master/LICENSE"
+    repo_url = "https://github.com/curl/curl-www/"
+    importer_name = "Curl Importer"
+    api_url = "https://curl.se/docs/vuln.json"
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:41]: @pombredanne @ambuj-1211 
I think the license_url should be this https://curl.se/docs/copyright.html
and the  `spdx_license_expression = curl`
we aren't getting the data from the Git repo

**Review by ziadhany — CHANGES_REQUESTED** [review]
LGTM, just a minor nit for your consideration

**Comment by TG1999** [comment]
@ambuj-1211 please resolve conflicts on this PR

**Comment by TG1999** [comment]
@ziadhany is this PR good to merge, if yes please merge this in. Thanks!

**Comment by ziadhany** [comment]
> @ziadhany is this PR good to merge, if yes please merge this in. Thanks!

Yes, I have reviewed this pull request several times and I believe we should merge it.

**Comment by TG1999** [comment]
@ambuj-1211 please see tests are failing

**Comment by ambuj-1211** [comment]
corrected the code now check @TG1999

File: vulnerabilities/importers/curl.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import logging
+from datetime import datetime
+from datetime import timezone
+from typing import Iterable
+from typing import Mapping
+
+import requests
```
Comment (ziadhany) [inline vulnerabilities/importers/curl.py:16]:
```suggestion

```

**Comment by ambuj-1211** [comment]
@TG1999 please have a look on the log file
[curlimporterlogs.txt](https://github.com/user-attachments/files/16947823/curlimporterlogs.txt)

**Comment by TG1999** [comment]
@ambuj-1211 thanks! merged!

**Pull request merged**